compiler-lib/src/semantics.rs

```diff
@@ -0,0 +1,162 @@
+                        let discr = ast::ClassMemberKindDiscriminants::from(&member.kind);
+                        self.classes_and_members.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
+                    "unexpected ast node passed to first pass: {:?}",
+                    x
+                ))
+            }
+        }
+    }
+}
+
+type ScopeDefs<T> = HashMap<Symbol, T>;
+
+struct Scope<T> {
+    parent: Option<Rc<Scope<T>>>,
+    definitions: ScopeDefs<T>,
+}
+
+impl<T> Scope<T> {
+    pub fn root() -> Rc<Self> {
+        Rc::new(Scope {
+            parent: None,
+            definitions: ScopeDefs::new(),
+        })
+    }
+    pub fn enter_scope(parent: Rc<Self>) -> Self {
+        Scope {
+            parent: Some(parent),
+            definitions: ScopeDefs::new(),
+        }
+    }
+    pub fn leave(self) -> Option<Rc<Self>> {
+        self.parent
+    }
+    pub fn define(&mut self, sym: &Symbol, val: T) -> Result<(), ()> {
+        use std::collections::hash_map::Entry;
+        let key = Rc::clone(sym);
+        match self.definitions.entry(key) {
+            Entry::Occupied(o) => return Err(()),
+            Entry::Vacant(e) => e.insert(val),
+        };
+        Ok(())
+    }
+}
+
+struct SecondPass;
+
+// TODO proper enum, but Type is definitely necessary
+type VarDef<'a, 'f> = &'a Spanned<'f, ast::Type>;
+struct TyDef; // TODO
+struct MethDef; // TODO
+
+impl SecondPass {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+        _classes_and_members: &ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        // TODO java.lang.System.out scope
+        // TODO ontop of that, scope of this package
+        let mut vardefs: Rc<Scope<VarDef<'a, 'f>>> = Scope::root();
+        let mut tydefs: Rc<Scope<TyDef>> = Scope::root();
+        let mut methdefs: Rc<Scope<MethDef>> = Scope::root();
+        Self::do_visit(&NodeKind::from(ast), context, vardefs, tydefs, methdefs)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        context: &context::Context<'_>,
+        vardefs: Rc<Scope<VarDef<'a, 'f>>>,
+        tydefs: Rc<Scope<TyDef>>,
+        methdefs: Rc<Scope<MethDef>>,
+    ) -> Result<(), Error> {
+        use self::NodeKind::*;
+        match n {
+            ClassMember(m) if m.kind.is_method() => {
+                // Assume types have been checked
+                // TODO enforce somewhere that main has String[] arg type
+                let params = m.kind.method_params().unwrap();
+                let mut vardefs = Scope::enter_scope(vardefs);
+                for p in &params.data {
+                    vardefs
+                        .define(&p.name, &p.ty)
```

Comment:
To realize this whitout have a full parameter list on `MainMethod`, make `VarDef` an enum of `Reserved` and `Binding(ast::Type)` and than treat main method specially here, adding a `vardefs.reserve(&mainmethod_symbol)`. I like that much better than carrying around the "main method params are a list of length 1 with the only parameter being an array of `String`" invariant